File: data-structure/js/lists/singly-linked-list-ht.ts
import { Node } from "./node";

class SinglyLinkedList {
  private head: Node | null = null;
  private tail: Node | null = null;

  appendOnHead(value: number) {
    const newNode = new Node(value);

    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }
  }

  appendOnTail(value: number) {
    const newNode = new Node(value);

    if (this.tail === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
  }

  show(): void {
    let current = this.head;

    if (current !== null) {
      while (current !== null) {
        console.log(current.value);
        current = current.next;
      }
    }
  }

  removeFromHead() {
    if (this.head === null) {
      console.log("The list is empty");
      return;
    }

    let aux = this.head;
    this.head = aux.next;

    return aux;
  }

  removeFromTail() {
    if (this.head === null) {
      console.log("The list is empty");
      return;
    }

    let previous: Node = this.head;
    if (previous.next === null) {
      this.head = null;
      this.tail = null;
    } else {
      while (previous.next!.value !== this.tail?.value) {
        previous = previous.next!;
      }
      previous.next = null;
      this.tail = previous;
    }

    return previous;
  }
}

const sll = new SinglyLinkedList();
sll.appendOnTail(1);
sll.appendOnTail(2);
sll.appendOnTail(3);
sll.appendOnHead(0);
sll.appendOnHead(-1);
sll.appendOnTail(6);
sll.show(); // -1 0 1 2 3 6
console.log(" ");
sll.removeFromHead();
sll.removeFromHead();
sll.removeFromTail();
sll.show(); // 1 2 3

File: data-structure/python/lists/singly-linked-list-ht.py
from node import Node

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def addOnHead(self, value):
        novo = Node(value)
        novo.next = self.head

        if self.head == None:
            self.tail = novo
        self.head = novo

    def addOnTail(self, value):
        novo = Node(value)

        if self.head == None:
            self.head = novo
        else:
            self.tail.next = novo
        self.tail = novo

    def show(self):
        current = self.head
        while current != None:
            current.getNode()
            current = current.next

    def removeFromHead(self):
        if self.head == None:
            print("The list is empty")
            return None
        
        temp = self.head
        if self.head.next == None:
            self.tail = None

        self.head = self.head.next
        return temp

    def removeFromTail(self):
        if self.head == None:
            print("The list is empty")
            return None

        removed_value = self.tail.value
        if self.tail == self.head:
            self.tail = None
            self.head = None
        else:
            current = self.head
            while current.next != self.tail:
                current = current.next
            
            current.next = None
            self.tail = current

        return removed_value

    
    def remove(self, value):
        if self.head == None:
            return None
        
        if self.head.value == value:
            self.head = self.head.next
            return self.head
        
        current = self.head
        previous = None
        while current != None:
            if current.value == value:
                previous.next = current.next
                return current
            previous = current
            current = current.next

        return None

    
    def find(self, value):
        if self.head == None:
            return None
        
        temp = self.head
        while temp != None:
            if temp.value == value:
                return temp
            temp = temp.next

        return None

sll = SinglyLinkedList()
sll.addOnHead(2)
sll.addOnHead(1)
sll.addOnTail(3)
sll.addOnTail(4)
sll.addOnTail(5)
sll.addOnHead(0)
sll.show()